### PR TITLE
Source containers input plugin

### PR DESCRIPTION
**Pull request by MartinBasti**
Maintainers will complete the following section:
- [ ] Commit messages are descriptive enough
- [ ] "Signed-off-by:" line is present in each commit
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request includes link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/plugins/input_osv3.py [hunk 1]
```diff
@@ -186,3 +187,68 @@ def run(self):
...
+
+    def run(self):
+        build_json = get_build_json()
+        image = os.environ['OUTPUT_IMAGE']
+        self.target_registry = os.environ.get('OUTPUT_REGISTRY', None)
+
+        arrangement_version = None
+        try:
+            user_params = os.environ['USER_PARAMS']
+            user_data = json.loads(user_params)  # TODO validate user params against schema
```
Comment (MartinBasti) [inline atomic_reactor/plugins/input_osv3.py]: I'd add validation in separate PR

File: atomic_reactor/plugins/input_osv3.py [hunk 2]
```diff
@@ -186,3 +187,68 @@ def run(self):
...
+
+        input_json = {
+            'image': image,
+            'openshift_build_selflink': build_json.get('metadata', {}).get('selfLink', None)
+        }
+        input_json.update(self.plugins_json)
+
+        self.log.debug("build json: %s", input_json)
+
+        # TODO remove koji plugins when koji is not defined
```
Comment (MartinBasti) [inline atomic_reactor/plugins/input_osv3.py]: I will make a note to koji-* plugins part to remove it when implemented/integrated

File: atomic_reactor/plugins/input_osv3.py [hunk 3]
```diff
@@ -186,3 +187,68 @@ def run(self):
...
+
+        return input_json
+
+    @classmethod
+    def is_autousable(cls):
+        return (
+            'OUTPUT_IMAGE' in os.environ and
+            'USER_PARAMS' in os.environ and
+            json.loads(os.environ['USER_PARAMS'])
+                .get('kind', '__unknown__') == 'source_containers_user_params'  # TODO use osbs-client constant
```
Comment (MartinBasti) [inline atomic_reactor/plugins/input_osv3.py]: :eyes: 

I'll try to create that PR soon in osbs-client, but it shouldn't prevent this PR from merging

File: atomic_reactor/plugins/input_osv3.py (hunk 2 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: Hmm, but without koji this feature is useless, should I fail here instead?

File: atomic_reactor/plugins/input_osv3.py [hunk 4]
```diff
@@ -186,3 +187,68 @@ def run(self):
...
+        arrangement_version = None
+        try:
+            user_params = os.environ['USER_PARAMS']
+            user_data = json.loads(user_params)  # TODO validate user params against schema
+            arrangement_version = user_data.get('arrangement_version', None)
+            self.plugins_json = get_plugins_with_user_data(user_params, user_data)
+            # if we get the USER_PARAMS, we'd better get the REACTOR_CONFIG too
+            reactor_config_map = os.environ['REACTOR_CONFIG']
+            self.reactor_env = read_yaml(reactor_config_map, 'schemas/config.json')
+        except KeyError:
```
Comment (rcerven) [inline atomic_reactor/plugins/input_osv3.py]: remove this, since we won't have that ATOMIC_REACTOR_PLUGINS variable defined for sources for sure

**Comment by rcerven** [comment]
code looks good

File: atomic_reactor/plugins/input_osv3.py (hunk 2 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: @rcerven @twaugh WDYT? Without enabled koji source-container builds are useless, shall I hard fail here? (Or somewhere else)

File: atomic_reactor/plugins/input_osv3.py (hunk 4 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: removed

File: atomic_reactor/plugins/input_osv3.py (hunk 2 above)
Comment (twaugh) [reply atomic_reactor/plugins/input_osv3.py]: Yes, we should hard fail.

**Comment by twaugh** [comment]
Looks good, added comment about the hard-fail-without-koji question.

**Comment by MartinBasti** [comment]
Actually this needs https://github.com/containerbuildsystem/osbs-client/pull/898 first merged into master

**Comment by rcerven** [comment]
looks good just tests

File: atomic_reactor/plugins/input_osv3.py [hunk 5]
```diff
@@ -186,3 +187,72 @@ def run(self):
...
+        # if we get the USER_PARAMS, we'd better get the REACTOR_CONFIG too
+        reactor_config_map = os.environ['REACTOR_CONFIG']
+        self.reactor_env = read_yaml(reactor_config_map, 'schemas/config.json')
+
+        if arrangement_version and arrangement_version <= 5:
+            raise ValueError('arrangement_version <= 5 is no longer supported')
+
+        self.plugins_json = json.loads(self.plugins_json)
+        # validate json before performing any changes
+        read_yaml(json.dumps(self.plugins_json), 'schemas/plugins.json')
```
Comment (twaugh) [inline atomic_reactor/plugins/input_osv3.py]: I know this is copied from elsewhere, but maybe this is better:
```
# validate json before performing any changes
read_yaml(self.plugins_json, 'schemas/plugins.json')
self.plugins_json = json.loads(self.plugins_json)
```

File: atomic_reactor/plugins/input_osv3.py [hunk 6]
```diff
@@ -186,3 +187,72 @@ def run(self):
...
+                "builds require enabled koji integration"
+            )
+
+    def run(self):
+        build_json = get_build_json()
+        image = os.environ['OUTPUT_IMAGE']
+        self.target_registry = os.environ.get('OUTPUT_REGISTRY', None)
+
+        user_params = os.environ['USER_PARAMS']
+        user_data = json.loads(user_params)  # TODO validate user params against schema
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: yes, but I need new schema, or even better move validation to osbs-client :question:

File: atomic_reactor/plugins/input_osv3.py [hunk 7]
```diff
@@ -186,3 +187,72 @@ def run(self):
...
+
+        return input_json
+
+    @classmethod
+    def is_autousable(cls):
+        return (
+            'OUTPUT_IMAGE' in os.environ and
+            'USER_PARAMS' in os.environ and
+            # TODO use osbs-client constant
+            json.loads(os.environ['USER_PARAMS'])
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: At this point, I don't know schema against it should be validated

File: atomic_reactor/plugins/input_osv3.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (twaugh) [reply atomic_reactor/plugins/input_osv3.py]: Oh, I see, ok.

File: atomic_reactor/plugins/input_osv3.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (twaugh) [reply atomic_reactor/plugins/input_osv3.py]: Gotcha.

File: atomic_reactor/plugins/input_osv3.py [hunk 8]
```diff
@@ -186,3 +187,72 @@ def run(self):
...
+                .get('kind', '__unknown__') == 'source_containers_user_params'
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: removed

File: atomic_reactor/plugins/input_osv3.py (hunk 5 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/input_osv3.py]: reworked

**Comment by MartinBasti** [comment]
I finished fight with tests, should be ready now

**Comment by MartinBasti** [comment]
No release notes, unfinished feature

**Pull request merged**